File: sdks/go/cmd/elydora/plugins/droid_contract.go
package plugins

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"math"
	"os/exec"
	"path/filepath"
	"runtime"
	"sort"
	"strings"
	"time"
)

const (
	droidAgentKey          = "droid"
	droidGuardScript       = "guard.js"
	droidAuditScript       = "hook.js"
	droidHookTimeout       = float64(10)
	droidOwnedFileMarker   = "// Managed by Elydora"
	droidPOSIXApostrophe   = `'"'"'`
	droidWindowsExitSuffix = "; exit $LASTEXITCODE"
)

var droidToolEvents = [...]string{"PreToolUse", "PostToolUse"}

type droidHookSettings map[string]any

type droidManagedRemoval struct {
	event          string
	groupIndex     int
	handlerIndexes []int
	removeGroup    bool
}

type droidRuntimeContract struct {
	agentID   string
	guardPath string
	auditPath string
}

type droidRegexEntry struct {
	Label   string `json:"label"`
	Pattern string `json:"pattern"`
}

func readDroidHookSettings(value any, label string) (droidHookSettings, error) {
	object, ok := value.(map[string]any)
	if !ok || object == nil {
		return nil, fmt.Errorf("%s must contain a JSON object", label)
	}
	hooks := make(droidHookSettings, len(object))
	for event, item := range object {
		groups, ok := item.([]any)
		if !ok {
			return nil, fmt.Errorf(`%s field %q must be an array`, label, event)
		}
		for groupIndex, group := range groups {
			if err := validateDroidGroup(group, label, event, groupIndex); err != nil {
				return nil, err
			}
		}
		hooks[event] = item
	}
	return hooks, nil
}

func validateDroidGroup(value any, label, event string, groupIndex int) error {
	location := fmt.Sprintf(`%s field %q[%d]`, label, event, groupIndex)
	group, ok := value.(map[string]any)
	if !ok || group == nil {
		return fmt.Errorf("%s must be an object", location)
	}
	if matcher, exists := group["matcher"]; exists {
		if _, ok := matcher.(string); !ok {
			return fmt.Errorf("%s matcher must be a string", location)
		}
	}
	if expression, exists := group["commandRegex"]; exists {
		if _, ok := expression.(string); !ok {
			return fmt.Errorf("%s commandRegex must be a string", location)
		}
	}
	handlers, ok := group["hooks"].([]any)
	if !ok {
		return fmt.Errorf("%s must contain a hooks array", location)
	}
	for handlerIndex, handler := range handlers {
		if err := validateDroidHandler(handler, location, handlerIndex); err != nil {
			return err
		}
	}
	return nil
}

func validateDroidHandler(value any, groupLabel string, handlerIndex int) error {
	location := fmt.Sprintf("%s.hooks[%d]", groupLabel, handlerIndex)
	handler, ok := value.(map[string]any)
	if !ok || handler == nil {
		return fmt.Errorf("%s must be an object", location)
	}
	if handler["type"] != "command" {
		return fmt.Errorf(`%s type must be "command"`, location)
	}
	command, ok := handler["command"].(string)
	if !ok || strings.TrimSpace(command) == "" {
		return fmt.Errorf("%s command must be a non-empty string", location)
	}
	if timeout, exists := handler["timeout"]; exists {
		number, ok := timeout.(float64)
		if !ok || math.IsNaN(number) || math.IsInf(number, 0) || number <= 0 {
			return fmt.Errorf("%s timeout must be a positive finite number", location)
		}
	}
	return nil
}

func validateDroidRegexes(settings ...droidHookSettings) error {
	entries := make([]droidRegexEntry, 0)
	for _, source := range settings {
		for event, value := range source {
			groups, _ := value.([]any)
			for index, groupValue := range groups {
				group := groupValue.(map[string]any)
				if matcher, ok := group["matcher"].(string); ok && matcher != "" && matcher != "*" {
					entries = append(entries, droidRegexEntry{
						Label:   fmt.Sprintf(`Factory Droid hooks field %q[%d] matcher`, event, index),
						Pattern: matcher,
					})
				}
				if expression, ok := group["commandRegex"].(string); ok {
					entries = append(entries, droidRegexEntry{
						Label:   fmt.Sprintf(`Factory Droid hooks field %q[%d] commandRegex`, event, index),
						Pattern: expression,
					})
				}
			}
		}
	}
	if len(entries) == 0 {
		return nil
	}
	nodePath, err := resolveNodeRuntime()
	if err != nil {
		return fmt.Errorf("Node.js runtime is required to validate Factory Droid hook expressions: %w", err)
	}
	payload, err := json.Marshal(entries)
	if err != nil {
		return fmt.Errorf("encode Factory Droid regular expressions: %w", err)
	}
	validator := `import fs from 'node:fs';
const entries = JSON.parse(fs.readFileSync(0, 'utf8'));
for (const entry of entries) {
  try { new RegExp(entry.pattern); }
  catch (error) {
    process.stderr.write(entry.label + ': ' + (error instanceof Error ? error.message : String(error)));
    process.exit(1);
  }
}`
	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()
	// #nosec G204 -- nodePath is resolved by exec.LookPath and every argument is fixed.
	command := exec.CommandContext(ctx, nodePath, "--input-type=module", "--eval", validator)
	command.Stdin = bytes.NewReader(payload)
	output, runErr := command.CombinedOutput()
	if ctx.Err() != nil {
		return fmt.Errorf("Factory Droid regular expression validation timed out: %w", ctx.Err())
	}
	if runErr != nil {
		message := strings.TrimSpace(string(output))
		if message == "" {
			message = runErr.Error()
		}
		return fmt.Errorf("Factory Droid matcher must be a valid JavaScript regular expression: %s", message)
	}
	return nil
}

func buildDroidCommand(nodePath, scriptPath string) string {
	if runtime.GOOS == "windows" {
		return "& " + quoteDroidPowerShellArgument(nodePath) + " " +
			quoteDroidPowerShellArgument(scriptPath) + droidWindowsExitSuffix
	}
	return quotePOSIXArgument(nodePath) + " " + quotePOSIXArgument(scriptPath)
}

func quoteDroidPowerShellArgument(value string) string {
	return "'" + strings.ReplaceAll(value, "'", "''") + "'"
}

func buildDroidGroup(nodePath, scriptPath string) map[string]any {
	return map[string]any{
		"matcher": "*",
		"hooks": []any{map[string]any{
			"type": "command", "command": buildDroidCommand(nodePath, scriptPath), "timeout": droidHookTimeout,
		}},
	}
}

func parseDroidCommand(command string, includeLegacy ...bool) (string, string, bool) {
	if runtime.GOOS != "windows" {
		return parseDroidPOSIXCommand(command)
	}
	executable, script, ok := parseDroidPowerShellCommand(command)
	if ok || len(includeLegacy) == 0 || !includeLegacy[0] {
		return executable, script, ok
	}
	return parseDroidLegacyWindowsCommand(command)
}

func parseDroidPOSIXCommand(command string) (string, string, bool) {
	executable, next, ok := readDroidPOSIXArgument(command, 0)
	if !ok || next >= len(command) || command[next] != ' ' {
		return "", "", false
	}
	script, end, ok := readDroidPOSIXArgument(command, next+1)
	return executable, script, ok && end == len(command) && executable != "" && script != ""
}

func parseDroidPowerShellCommand(command string) (string, string, bool) {
	if !strings.HasPrefix(command, "& ") {
		return "", "", false
	}
	executable, next, ok := readDroidPowerShellArgument(command, 2)
	if !ok || next >= len(command) || command[next] != ' ' {
		return "", "", false
	}
	script, end, ok := readDroidPowerShellArgument(command, next+1)
	return executable, script, ok && command[end:] == droidWindowsExitSuffix && executable != "" && script != ""
}

func parseDroidLegacyWindowsCommand(command string) (string, string, bool) {
	executable, next, ok := readDroidLegacyWindowsArgument(command, 0)
	if !ok || next >= len(command) || command[next] != ' ' {
		return "", "", false
	}
	script, end, ok := readDroidLegacyWindowsArgument(command, next+1)
	return executable, script, ok && end == len(command) && executable != "" && script != ""
}

func readDroidPowerShellArgument(command string, start int) (string, int, bool) {
	if start >= len(command) || command[start] != '\'' {
		return "", start, false
	}
	var value strings.Builder
	for index := start + 1; index < len(command); index++ {
		if command[index] != '\'' {
			value.WriteByte(command[index])
			continue
		}
		if index+1 < len(command) && command[index+1] == '\'' {
			value.WriteByte('\'')
			index++
			continue
		}
		return value.String(), index + 1, true
	}
	return "", start, false
}

func readDroidLegacyWindowsArgument(command string, start int) (string, int, bool) {
	if start >= len(command) || command[start] != '"' {
		return "", start, false
	}
	end := strings.IndexByte(command[start+1:], '"')
	if end < 0 {
		return "", start, false
	}
	end += start + 1
	value := command[start+1 : end]
	return value, end + 1, value != "" && !strings.ContainsAny(value, "\r\n")
}

func readDroidPOSIXArgument(command string, start int) (string, int, bool) {
	if start >= len(command) || command[start] != '\'' {
		return "", start, false
	}
	var value strings.Builder
	for index := start + 1; index < len(command); {
		if strings.HasPrefix(command[index:], droidPOSIXApostrophe) {
			value.WriteByte('\'')
			index += len(droidPOSIXApostrophe)
			continue
		}
		if command[index] == '\'' {
			return value.String(), index + 1, true
		}
		value.WriteByte(command[index])
		index++
	}
	return "", start, false
}

func sameDroidPath(left, right string) bool {
	left = filepath.Clean(left)
	right = filepath.Clean(right)
	if absolute, err := filepath.Abs(left); err == nil {
		left = absolute
	}
	if absolute, err := filepath.Abs(right); err == nil {
		right = absolute
	}
	if runtime.GOOS == "windows" {
		return strings.EqualFold(left, right)
	}
	return left == right
}

func sameDroidAgentID(left, right string) bool {
	if runtime.GOOS == "windows" {
		return strings.EqualFold(left, right)
	}
	return left == right
}

func droidRuntimeRoot() (string, error) {
	return AgentRuntimeRoot()
}

func managedDroidAgentID(
	handler map[string]any,
	scriptName, runtimeRoot string,
	includeLegacy ...bool,
) (string, bool) {
	if len(handler) != 3 || handler["type"] != "command" || handler["timeout"] != droidHookTimeout {
		return "", false
	}
	command, ok := handler["command"].(string)
	if !ok {
		return "", false
	}
	executable, scriptPath, ok := parseDroidCommand(command, includeLegacy...)
	if !ok || !filepath.IsAbs(executable) || !filepath.IsAbs(scriptPath) ||
		!isDroidNodeExecutable(executable) || !sameDroidFileName(filepath.Base(scriptPath), scriptName) {
		return "", false
	}
	agentDirectory := filepath.Dir(scriptPath)
	if !sameDroidPath(filepath.Dir(agentDirectory), runtimeRoot) {
		return "", false
	}
	agentID := filepath.Base(agentDirectory)
	return agentID, agentID != "" && agentID != "." && agentID != ".."
}

func isDroidNodeExecutable(path string) bool {
	name := filepath.Base(path)
	return name == "node" || strings.EqualFold(name, "node.exe")
}

func sameDroidFileName(left, right string) bool {
	if runtime.GOOS == "windows" {
		return strings.EqualFold(left, right)
	}
	return left == right
}

func managedDroidRemovals(settings droidHookSettings, agentID, runtimeRoot string) []droidManagedRemoval {
	removals := make([]droidManagedRemoval, 0)
	for _, contract := range []struct{ event, script string }{
		{"PreToolUse", droidGuardScript}, {"PostToolUse", droidAuditScript},
	} {
		groups, _ := settings[contract.event].([]any)
		for groupIndex, value := range groups {
			group := value.(map[string]any)
			handlers := group["hooks"].([]any)
			indexes := make([]int, 0)
			for handlerIndex, handlerValue := range handlers {
				managedID, managed := managedDroidAgentID(
					handlerValue.(map[string]any), contract.script, runtimeRoot, true,
				)
				if managed && (agentID == "" || sameDroidAgentID(managedID, agentID)) {
					indexes = append(indexes, handlerIndex)
				}
			}
			if len(indexes) == 0 {
				continue
			}
			_, hasMatcher := group["matcher"]
			_, hasHooks := group["hooks"]
			exactGroup := len(group) == 2 && hasMatcher && hasHooks &&
				group["matcher"] == "*" && len(indexes) == len(handlers)
			removals = append(removals, droidManagedRemoval{
				contract.event, groupIndex, indexes, exactGroup,
			})
		}
	}
	return removals
}

func droidRuntimeContracts(settings droidHookSettings, runtimeRoot string) []droidRuntimeContract {
	guards := managedDroidIDs(settings, "PreToolUse", droidGuardScript, runtimeRoot)
	audits := managedDroidIDs(settings, "PostToolUse", droidAuditScript, runtimeRoot)
	guardIDs := make([]string, 0, len(guards))
	for agentID := range guards {
		guardIDs = append(guardIDs, agentID)
	}
	sort.Strings(guardIDs)
	contracts := make([]droidRuntimeContract, 0, len(guardIDs))
	for _, guardID := range guardIDs {
		for auditID := range audits {
			if !sameDroidAgentID(guardID, auditID) {
				continue
			}
			root := filepath.Join(runtimeRoot, guardID)
			contracts = append(contracts, droidRuntimeContract{
				guardID,
				filepath.Join(root, droidGuardScript),
				filepath.Join(root, droidAuditScript),
			})
			break
		}
	}
	return contracts
}

func managedDroidIDs(settings droidHookSettings, event, script, runtimeRoot string) map[string]bool {
	ids := map[string]bool{}
	groups, _ := settings[event].([]any)
	for _, groupValue := range groups {
		group := groupValue.(map[string]any)
		handlers := group["hooks"].([]any)
		if len(group) != 2 || group["matcher"] != "*" || len(handlers) != 1 {
			continue
		}
		if agentID, ok := managedDroidAgentID(handlers[0].(map[string]any), script, runtimeRoot); ok {
			ids[agentID] = true
		}
	}
	return ids
}

File: sdks/node/src/plugins/droid-contract.ts
import os from 'node:os';
import path from 'node:path';
import type { JsonObject } from './strict-json.js';

export const AGENT_KEY = 'droid';
export const GUARD_SCRIPT = 'guard.js';
export const AUDIT_SCRIPT = 'hook.js';
export const HOOK_TIMEOUT_SECONDS = 10;
export const TOOL_EVENTS = ['PreToolUse', 'PostToolUse'] as const;

const HANDLER_KEYS = ['command', 'timeout', 'type'];
const GROUP_KEYS = ['hooks', 'matcher'];
const WINDOWS_EXIT_SUFFIX = '; exit $LASTEXITCODE';

export type ToolEvent = typeof TOOL_EVENTS[number];

export interface DroidHandler extends JsonObject {
  readonly type: 'command';
  readonly command: string;
  readonly timeout?: number;
}

export interface DroidGroup extends JsonObject {
  readonly matcher?: string;
  readonly commandRegex?: string;
  readonly hooks: DroidHandler[];
}

export interface DroidHookMap extends JsonObject {
  readonly PreToolUse?: DroidGroup[];
  readonly PostToolUse?: DroidGroup[];
}

export interface RuntimeContract {
  readonly agentId: string;
  readonly guardPath: string;
  readonly auditPath: string;
}

export interface ManagedRemoval {
  readonly event: ToolEvent;
  readonly groupIndex: number;
  readonly handlerIndexes: number[];
  readonly removeGroup: boolean;
}

interface ParsedArgument {
  readonly value: string;
  readonly next: number;
}

function quotePosix(value: string): string {
  return `'${value.replaceAll("'", `'"'"'`)}'`;
}

function quotePowerShell(value: string): string {
  return `'${value.replaceAll("'", "''")}'`;
}

export function buildCommand(scriptPath: string): string {
  if (!path.isAbsolute(process.execPath) || !path.isAbsolute(scriptPath)) {
    throw new Error('Factory Droid hook commands require absolute executable and script paths');
  }
  return process.platform === 'win32'
    ? `& ${quotePowerShell(process.execPath)} ${quotePowerShell(scriptPath)}${WINDOWS_EXIT_SUFFIX}`
    : `${quotePosix(process.execPath)} ${quotePosix(scriptPath)}`;
}

export function buildGroup(scriptPath: string): DroidGroup {
  return {
    matcher: '*',
    hooks: [{
      type: 'command',
      command: buildCommand(scriptPath),
      timeout: HOOK_TIMEOUT_SECONDS,
    }],
  };
}

function validateRegex(value: string, label: string, allowWildcard: boolean): void {
  if (allowWildcard && (value === '*' || value === '')) return;
  try {
    new RegExp(value);
  } catch (error) {
    const message = error instanceof Error ? error.message : String(error);
    throw new Error(`${label} must be a valid regular expression: ${message}`);
  }
}

function validateHandler(
  value: unknown,
  label: string,
  groupIndex: number,
  handlerIndex: number,
): DroidHandler {
  const location = `${label}[${groupIndex}].hooks[${handlerIndex}]`;
  if (typeof value !== 'object' || value === null || Array.isArray(value)) {
    throw new Error(`${location} must be an object`);
  }
  const handler = value as JsonObject;
  if (handler.type !== 'command') throw new Error(`${location} type must be "command"`);
  if (typeof handler.command !== 'string' || handler.command.trim().length === 0) {
    throw new Error(`${location} command must be a non-empty string`);
  }
  if (handler.timeout !== undefined
    && (typeof handler.timeout !== 'number'
      || !Number.isFinite(handler.timeout)
      || handler.timeout <= 0)) {
    throw new Error(`${location} timeout must be a positive finite number`);
  }
  return handler as DroidHandler;
}

function validateGroup(value: unknown, label: string, groupIndex: number): DroidGroup {
  const location = `${label}[${groupIndex}]`;
  if (typeof value !== 'object' || value === null || Array.isArray(value)) {
    throw new Error(`${location} must be an object`);
  }
  const group = value as JsonObject;
  if (group.matcher !== undefined) {
    if (typeof group.matcher !== 'string') throw new Error(`${location} matcher must be a string`);
    validateRegex(group.matcher, `${location} matcher`, true);
  }
  if (group.commandRegex !== undefined) {
    if (typeof group.commandRegex !== 'string') {
      throw new Error(`${location} commandRegex must be a string`);
    }
    validateRegex(group.commandRegex, `${location} commandRegex`, false);
  }
  if (!Array.isArray(group.hooks)) throw new Error(`${location} must contain a hooks array`);
  return {
    ...group,
    hooks: group.hooks.map(
      (handler, handlerIndex) => validateHandler(handler, label, groupIndex, handlerIndex),
    ),
  } as DroidGroup;
}

export function readHookMap(value: unknown, label: string): DroidHookMap {
  if (typeof value !== 'object' || value === null || Array.isArray(value)) {
    throw new Error(`${label} must contain a JSON object`);
  }
  const hooks: DroidHookMap = {};
  for (const [event, groups] of Object.entries(value)) {
    if (!Array.isArray(groups)) throw new Error(`${label} field "${event}" must be an array`);
    hooks[event] = groups.map(
      (group, groupIndex) => validateGroup(group, `${label} field "${event}"`, groupIndex),
    );
  }
  return hooks;
}

function readPowerShellArgument(command: string, start: number): ParsedArgument | undefined {
  if (command[start] !== "'") return undefined;
  let value = '';
  for (let index = start + 1; index < command.length; index += 1) {
    if (command[index] !== "'") {
      value += command[index];
      continue;
    }
    if (command[index + 1] === "'") {
      value += "'";
      index += 1;
      continue;
    }
    return { value, next: index + 1 };
  }
  return undefined;
}

const POSIX_APOSTROPHE = `'"'"'`;

function readPosixArgument(command: string, start: number): ParsedArgument | undefined {
  if (command[start] !== "'") return undefined;
  let value = '';
  for (let index = start + 1; index < command.length;) {
    if (command.startsWith(POSIX_APOSTROPHE, index)) {
      value += "'";
      index += POSIX_APOSTROPHE.length;
      continue;
    }
    if (command[index] === "'") return { value, next: index + 1 };
    value += command[index];
    index += 1;
  }
  return undefined;
}

function parseTwoArguments(
  command: string,
  start: number,
  readArgument: (command: string, start: number) => ParsedArgument | undefined,
): readonly [string, string] | undefined {
  const executable = readArgument(command, start);
  if (!executable || command[executable.next] !== ' ') return undefined;
  const script = readArgument(command, executable.next + 1);
  if (!script || script.next !== command.length || !executable.value || !script.value) return undefined;
  return [executable.value, script.value];
}

function parseLegacyWindowsCommand(command: string): readonly [string, string] | undefined {
  const match = /^"([^"\r\n]+)" "([^"\r\n]+)"$/.exec(command);
  return match ? [match[1], match[2]] : undefined;
}

function parsePowerShellCommand(command: string): readonly [string, string] | undefined {
  if (!command.startsWith('& ')) return undefined;
  const executable = readPowerShellArgument(command, 2);
  if (!executable || command[executable.next] !== ' ') return undefined;
  const script = readPowerShellArgument(command, executable.next + 1);
  if (!script
    || command.slice(script.next) !== WINDOWS_EXIT_SUFFIX
    || !executable.value
    || !script.value) return undefined;
  return [executable.value, script.value];
}

function parseGeneratedCommand(
  command: string,
  includeLegacy: boolean,
): readonly [string, string] | undefined {
  if (process.platform !== 'win32') return parseTwoArguments(command, 0, readPosixArgument);
  const current = parsePowerShellCommand(command);
  return current ?? (includeLegacy ? parseLegacyWindowsCommand(command) : undefined);
}

export function samePath(left: string, right: string): boolean {
  const normalizedLeft = path.resolve(left);
  const normalizedRight = path.resolve(right);
  return process.platform === 'win32'
    ? normalizedLeft.toLowerCase() === normalizedRight.toLowerCase()
    : normalizedLeft === normalizedRight;
}

export function sameAgentId(left: string, right: string): boolean {
  return process.platform === 'win32' ? left.toLowerCase() === right.toLowerCase() : left === right;
}

export function managedAgentId(
  handler: DroidHandler,
  scriptName: string,
  includeLegacy = false,
): string | undefined {
  if (Object.keys(handler).sort().join('\0') !== HANDLER_KEYS.join('\0')
    || handler.type !== 'command'
    || handler.timeout !== HOOK_TIMEOUT_SECONDS) return undefined;
  const parsed = parseGeneratedCommand(handler.command, includeLegacy);
  if (!parsed || !samePath(parsed[0], process.execPath)) return undefined;
  const scriptPath = parsed[1];
  if (path.basename(scriptPath) !== scriptName) return undefined;
  const agentDirectory = path.dirname(scriptPath);
  if (!samePath(path.dirname(agentDirectory), path.join(os.homedir(), '.elydora'))) return undefined;
  const agentId = path.basename(agentDirectory);
  return agentId && agentId !== '.' && agentId !== '..' ? agentId : undefined;
}

function exactOwnedGroup(group: DroidGroup, managedIndexes: number[]): boolean {
  return Object.keys(group).sort().join('\0') === GROUP_KEYS.join('\0')
    && group.matcher === '*'
    && managedIndexes.length > 0
    && managedIndexes.length === group.hooks.length;
}

export function managedRemovals(hooks: DroidHookMap, agentId?: string): ManagedRemoval[] {
  const removals: ManagedRemoval[] = [];
  for (const [event, scriptName] of [
    ['PreToolUse', GUARD_SCRIPT],
    ['PostToolUse', AUDIT_SCRIPT],
  ] as const) {
    const groups = hooks[event] ?? [];
    groups.forEach((group, groupIndex) => {
      const handlerIndexes = group.hooks.flatMap((handler, handlerIndex) => {
        const managedId = managedAgentId(handler, scriptName, true);
        return managedId && (agentId === undefined || sameAgentId(managedId, agentId))
          ? [handlerIndex]
          : [];
      });
      if (handlerIndexes.length > 0) {
        removals.push({
          event,
          groupIndex,
          handlerIndexes,
          removeGroup: exactOwnedGroup(group, handlerIndexes),
        });
      }
    });
  }
  return removals;
}

function configuredIds(groups: DroidGroup[], scriptName: string): Set<string> {
  const ids = new Set<string>();
  for (const group of groups) {
    if (Object.keys(group).sort().join('\0') !== GROUP_KEYS.join('\0')
      || group.matcher !== '*'
      || group.hooks.length !== 1) continue;
    const agentId = managedAgentId(group.hooks[0], scriptName);
    if (agentId) ids.add(agentId);
  }
  return ids;
}

export function runtimeContracts(hooks: DroidHookMap): RuntimeContract[] {
  const guards = configuredIds(hooks.PreToolUse ?? [], GUARD_SCRIPT);
  const audits = configuredIds(hooks.PostToolUse ?? [], AUDIT_SCRIPT);
  const root = path.join(os.homedir(), '.elydora');
  return [...guards]
    .filter((agentId) => [...audits].some((auditId) => sameAgentId(auditId, agentId)))
    .map((agentId) => ({
      agentId,
      guardPath: path.join(root, agentId, GUARD_SCRIPT),
      auditPath: path.join(root, agentId, AUDIT_SCRIPT),
    }));
}
